pallets/migrations/src/lib.rs

```diff
@@ -0,0 +1,209 @@
+
+		/// on_runtime_upgrade is expected to be called exactly once after a runtime upgrade.
+		/// We use this as a chance to flag that we are now in upgrade-mode and begin our
+		/// migrations.
+		/// 
+		/// In the event that a migration is expected to take more than one block, ongoing migration
+		/// work could continue from block-to-block in this pallet's on_initialize function.
+		fn on_runtime_upgrade() -> Weight {
+			log::warn!("Performing on_runtime_upgrade");
+
+			// start by flagging that we are not fully upgraded
+			<FullyUpgraded<T>>::put(false);
+
+			let mut weight: Weight = 0u64.into();
+
+			weight += process_runtime_upgrades::<T>();
+
+			weight.into()
+		}
+	}
+
+	#[pallet::storage]
+	#[pallet::getter(fn is_fully_upgraded)]
+	/// True if all required migrations have completed
+	type FullyUpgraded<T: Config> = StorageValue<_, bool, ValueQuery>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn migration_state)]
+	/// MigrationState tracks the progress of a migration.
+	type MigrationState<T: Config> = StorageMap<
+		_,
+		Twox64Concat,
+		String,
+		Perbill,
+		OptionQuery, // TODO: what is this...?
+	>;
+
+	#[pallet::genesis_config]
+	pub struct GenesisConfig<T: Config> {
+		pub completed_migrations: Vec<String>,
+		pub dummy: PhantomData<T> // TODO: 
+	}
+
+	#[cfg(feature = "std")]
+	impl<T: Config> Default for GenesisConfig<T> {
+		fn default() -> Self {
+			Self {
+				completed_migrations: vec![],
+				dummy: PhantomData,
+			}
+		}
+	}
+
+	#[pallet::genesis_build]
+	impl<T: Config> GenesisBuild<T> for GenesisConfig<T> {
+		fn build(&self) {
+			for migration_name in &self.completed_migrations {
+				<MigrationState<T>>::insert(migration_name, Perbill::one());
+			}
+		}
+	}
+
+	fn process_runtime_upgrades<T: Config>() -> Weight {
+		log::info!("stepping runtime upgrade");
+
+		// TODO: query proper value or make configurable
+		let available_weight = 500_000_000_000u64.into();
+		let mut weight: Weight = 0u64.into();
+		let mut done: bool = true;
+
+		for migration in &MIGRATIONS {
+
+			// let migration_name = migration.friendly_name();
+			let migration_name = migration.friendly_name();
+			log::trace!("evaluating migration {}", migration_name);
+
+			let migration_state = <MigrationState<T>>::get(migration_name)
+				.unwrap_or(Perbill::zero());
+
+			if migration_state < Perbill::one() {
+
+				let available_for_step = available_weight - weight;
+				log::trace!("stepping migration {}, prev: {:?}, avail weight: {}",
+					migration_name, migration_state, available_for_step);
+
+				// perform a step of this migration
+				let (updated_progress, consumed_weight)
+					= migration.step(migration_state, available_for_step);
+
+				weight += consumed_weight;
+				if weight > available_weight {
+					// TODO: the intent here is to complain obnoxiously so that this is caught
+					// during development. In production, this should probably be tolerated because
+					// failing is catastrophic.
```

Comment:
so you could `break` out of the migrations loop here and return the `weight` which is really the `total_consumed_weight`?